gui/wxpython/main_window/notebook.py

```diff
@@ -0,0 +1,141 @@
+        client_size = self.parent.GetClientSize()
+        self.SetPosition(wx.Point(client_size.x, client_size.y))
+        self.SetSize(430, 200)
+
+        self.SetArtProvider(SimpleTabArt())
+
+        # bindings
+        self.Bind(
+            aui.EVT_AUINOTEBOOK_PAGE_CHANGED,
+            lambda evt: self.GetCurrentPage().onFocus.emit(),
+        )
+        self.Bind(aui.EVT_AUINOTEBOOK_PAGE_CLOSE, self.OnClose)
+
+    def UndockMapDisplay(self, page):
+        """Undock active map display to independent MapFrame object"""
+        idx = self.GetPageIndex(page)
+        text = self.GetPageText(idx)
+        self.RemovePage(idx)
+        fr = MapPageFrame(parent=self.parent, mapdisplay=page, title=text)
+        page.Reparent(fr)
+        page.SetDockingCallback(self.DockMapDisplay)
+        fr.sizer.Add(page, proportion=1, flag=wx.EXPAND)
+        fr.Show()
+        page.Show()
+        page.onFocus.emit()
+
+    def DockMapDisplay(self, page):
+        """Dock independent MapFrame object back to Aui.Notebook"""
+        fr = page.GetParent()
+        page.Reparent(self)
+        page.SetDockingCallback(self.UndockMapDisplay)
+        self.AddPage(page, fr.GetTitle())
+        fr.closeFrameNoEvent()
+
+    def AddPage(self, *args, **kwargs):
+        """Overrides Aui.Notebook AddPage method.
+        Adds page to notebook and make it current"""
+        super().AddPage(*args, **kwargs)
+        self.SetSelection(self.GetPageCount() - 1)
+
+    def SetSelectionToPage(self, page):
+        """Overrides Aui.Notebook SetSelectionToPage method.
+        Decides whether to set selection to a MapNotebook page
+        or an undocked independent frame"""
+        try:
+            super().SetSelection(self.GetPageIndex(page))
+        except Exception:
+            pass
+
+        if not page.IsDocked():
+            wx.CallLater(500, page.GetParent().Raise)
+
+    def DeletePage(self, page):
+        """Overrides Aui.Notebook DeletePage method.
+        Decides whether to destroy a MapNotebook page
+        or an undocked independent frame"""
+        try:
+            super().DeletePage(self.GetPageIndex(page))
+        except Exception:
+            pass
+
+        if not page.IsDocked():
+            page.Destroy()
+
+    def SetPageText(self, page, text):
+        """Overrides Aui.Notebook SetPageText method.
+        Decides whether sets title to MapNotebook page
+        or an undocked independent frame"""
+        try:
+            super().SetPageText(page_idx=self.GetPageIndex(page), text=text)
+        except Exception:
+            pass
+
+        if not page.IsDocked():
+            frame = page.GetParent()
+            frame.SetTitle(text)
+            wx.CallLater(500, frame.Raise)
+
+    def OnClose(self, event):
+        """Page of map notebook is being closed"""
+        display = self.GetCurrentPage()
+        display.OnCloseWindow(event=None, askIfSaveWorkspace=True)
+        event.Veto()
```

Comment:
event.veto() ??

Reply:
it just a copy paste of the code:

   def OnMapNotebookClose(self, event):
        """Page of map notebook is being closed"""
        display = self.GetMapDisplay(onlyCurrent=True)
        display.OnCloseWindow(event=None, askIfSaveWorkspace=True)
        event.Veto()
I tested it and it does not work without event.Veto()